src/parser/constructs.rs

```diff
@@ -339,23 +339,24 @@ pub fn block(input: &str) -> ParseResult<&str, Block> {
     inner_block(input)
 }
 
-/// inner_block = expr ( ';' expr )* '}'
-///             | '}'
+/// inner_block = '}'
+///             | expr '}'                  (* The only case where block is an expr *)
+///             | expr ';' next inner_block
 fn inner_block(input: &str) -> ParseResult<&str, Block> {
-    let mut block = Block::new();
     if let Ok((input, _)) = Token::right_curly_bracket(input) {
-        return Ok((input, block));
+        return Ok((input, Block::new()));
     }
 
-    let (mut input, inst) = expr(input)?;
-    block.add_instruction(inst);
-    while let Ok((new_input, _)) = Token::semicolon(input) {
-        let (new_input, inst) = expr(new_input)?;
+    let (input, inst) = expr(input)?;
+    if let Ok((input, _)) = Token::right_curly_bracket(input) {
+        let mut block = Block::new();
         block.add_instruction(inst);
-        input = new_input;
+        block.set_statement(false);
+        return Ok((input, block));
     }
-    let (input, _) = Token::right_curly_bracket(input)?;
 
+    let (input, mut block) = preceded(Token::semicolon, preceded(nom_next, inner_block))(input)?;
+    block.push_front_instruction(inst);
```

Comment:
Why push_front here?

Reply:
Explained in private: the function is recursive, so we need to push front the expression we had already parsed.